runtime/service.go

```diff
@@ -171,96 +183,61 @@ func NewService(shimCtx context.Context, id string, remotePublisher shim.Publish
-	return vm.Dir(filepath.Join(varRunDir, s.namespace, s.vmID))
-}
-
-func (s *service) newShim(namespace string, containerdBinary, containerdAddress string, shimSocket *net.UnixListener) (*exec.Cmd, error) {
-	self, err := os.Executable()
-	if err != nil {
-		return nil, err
-	}
-
-	args := []string{
-		"-namespace", namespace,
-		"-address", containerdAddress,
-		"-publish-binary", containerdBinary,
-	}
-
-	if s.config.Debug {
-		args = append(args, "-debug")
-	}
 
-	cmd := exec.Command(self, args...)
-
-	cmd.Dir = s.vmDir().RootPath()
-
-	cmd.Env = append(os.Environ(),
-		fmt.Sprintf("%s=%s", vmIDEnvVarKey, s.vmID))
-
-	shimSocketFile, err := shimSocket.File()
-	if err != nil {
-		return nil, err
-	}
-	cmd.ExtraFiles = append(cmd.ExtraFiles, shimSocketFile)
-
-	cmd.SysProcAttr = &syscall.SysProcAttr{
-		Setpgid: true,
+		vmReady: make(chan struct{}),
 	}
 
-	err = cmd.Start()
-	if err != nil {
-		return nil, err
-	}
+	// TODO we have to create separate listeners for the fccontrol service and shim service because
+	// containerd does not currently expose the shim server for us to register the fccontrol service with too.
+	// This is likely addressable through some relatively small upstream contributions; the following is a stop-gap
+	// solution until that time.
+	fcSocketFDEnvVal := os.Getenv(internal.FCSocketFDEnvKey)
+	if fcSocketFDEnvVal != "" {
+		socketFD, err := strconv.Atoi(fcSocketFDEnvVal)
+		if err != nil {
+			return nil, err
+		}
 
-	// make sure to wait after start
-	go func() {
-		s.logger.WithField("vmID", s.vmID).Debug("waiting on shim process")
-		waitErr := cmd.Wait()
-		s.logger.WithError(waitErr).WithField("vmID", s.vmID).Debug("completed waiting on shim process")
-	}()
+		fccontrolTtrpc.RegisterFirecrackerService(fcServer, s)
+		fcListener, err := net.FileListener(os.NewFile(uintptr(socketFD), "fcsocket"))
+		if err != nil {
+			return nil, err
+		}
 
-	err = shim.SetScore(cmd.Process.Pid)
-	if err != nil {
-		s.logger.WithError(err).WithField("vmID", s.vmID).Error("failed to set OOM score on shim")
-		return nil, errors.Wrap(err, "failed to set OOM Score on shim")
+		go func() {
+			defer fcListener.Close()
+			err := fcServer.Serve(shimCtx, fcListener)
+			if err != nil && !strings.Contains(err.Error(), "use of closed network connection") {
+				s.logger.WithError(err).Error("fccontrol ttrpc server error")
+			}
+		}()
 	}
 
-	return cmd, nil
-}
-
-func isEADDRINUSE(err error) bool {
-	return err != nil && strings.Contains(err.Error(), "address already in use")
+	return s, nil
 }
 
 func (s *service) StartShim(shimCtx context.Context, containerID, containerdBinary, containerdAddress string) (string, error) {
-	s.logger.WithField("id", containerID).Debug("StartShim")
-
-	namespace, err := namespaces.NamespaceRequired(shimCtx)
-	if err != nil {
-		return "", errors.Wrap(err, "failed to extract namespace from context")
-	}
+	logrus.SetOutput(os.Stderr) // containerd reads the socket address from stdout only during shim start
```

Comment:
That's what I wanted to do (to just always use Stderr), but that results in another issue that requires going down a rabbit hole to explain fully.

If you always use Stderr for the logger here, then the plugin will have to assume that anything written to Stderr is already pre-formatted via logrus and thus should be passed through to containerd logs unmodified ([see the 4 lines here for more context](https://github.com/sipsma/firecracker-containerd/blob/fccontrol/firecracker-control/local.go#L291)).

That becomes a problem if we hit [this line in containerd's shim code](https://github.com/containerd/containerd/blob/master/runtime/v2/shim/shim.go#L147), which occurs if the shim is exiting with an error. That line prints to stderr directly without any logrus formatting, so the end result is your containerd logs have this random line with no timestamp or context.

The solution here works because the StartShim method is never hit by the plugin; the plugin only ever starts up the actual runtime shim, it doesn't invoke the start shim or delete shim routines.

Obviously it's tempting to just update containerd to print that line with the proper logrus formatting. However, that would result in the actual error object containerd eventually constructs server-side having the full logrus formatting, which would probably be confusing when returned to clients.

Overall, the fundamental reason this is all so messy right now is that we can't rely on containerd's normal way of providing shim logging due to [the issues I mention in this commit (part of this PR)](https://github.com/sipsma/firecracker-containerd/commit/d27ed36ceb1ed81ca462d52022c6b85edd4316ab).

I think the best long-term way of addressing this would be to figure out how to update containerd to not tie a shim log FIFO's existence to the existence of a bundle directory. That's a fairly sizable endeavor though separate from the current effort.

All that being said, I did realize while going through this all in my head again that 1. the comment I have is slightly inaccurate (containerd reads `CombinedOutput`, which is both stdout and stderr, not just stdout) and 2. I need to remove the `.Debug` log call below as, if debug is on, that will result in the debug log line being interpreted as part of the address string incorrectly.

I think what I'll do for the moment is change the StartShim logger to not use `os.Stderr` and instead actually use the log FIFO provided by containerd, which is okay because StartShim will always be invoked inside a bundle dir (an assumption that doesn't hold for the other parts of our code). That's probably the "minimal kludge" solution for now until we can look into the better long-term solution I mentioned above.